File: behavioral/state.go
package main

import "fmt"

type State interface{ DoAction(*Context) }
type Context struct{ state State }

func (c *Context) SetState(s State) { c.state = s }

type StartState struct{}

func (StartState) DoAction(c *Context) { fmt.Println("Start"); c.SetState(StopState{}) }

type StopState struct{}

func (StopState) DoAction(c *Context) { fmt.Println("Stop") }

func main() {
	c := &Context{state: StartState{}}
	c.state.DoAction(c)
	c.state.DoAction(c)
}

File: behavioral/state.php
<?php
interface State { public function doAction(Context $c): void; }
class Context { public State $state; public function __construct(State $s){ $this->state = $s; } }
class StartState implements State {
    public function doAction(Context $c): void { echo "Start\n"; $c->state = new StopState(); }
}
class StopState implements State {
    public function doAction(Context $c): void { echo "Stop\n"; }
}

$c = new Context(new StartState());
$c->state->doAction($c); // Start -> switches to Stop
$c->state->doAction($c); // Stop
